### PR TITLE
Customize layer assignment problem

### PR DESCRIPTION
**Pull request by davide-f**
This commit enables to slightly change the layer assignment problem highlighted in #16 ; yet, for simplicity, nodes cannot be moved beyond layers containing their children nodes.
This change enables the creation of this map:
image
The original plot had nodes 6 at layer 3 and node 8 at layer 4, instead.

### CODE REVIEW

File: src/layering.jl [hunk 1]
```diff
@@ -1,9 +1,36 @@
...
+        opt_ids_by_layers = sortperm(collect(values(opt_layer_assign)), rev=true)
+        key_opt = key_opt[opt_ids_by_layers]
+        values_opt = values_opt[opt_ids_by_layers]
+
+        # change layers if needed
+        for (k, l) in zip(key_opt, values_opt)
+            #current layer of node
+            curr_layer = findfirst(k .∈ layer_groups)
+            if curr_layer >= l
+                @warn "Ignored opt_layer_assign for node $k; dists[k] ($(dists[k])) > l ($l)"
```
Comment (davide-f) [inline src/layering.jl:25]: In the case the user specify a layer lower than the current one, the optional value is ignored with a warning

**Review by oxinabox — COMMENTED** [review]
Sorry again for the delay in reviewing.
This seems really complex.
I think it might be simpler if rather than post-processing the `layer_groups`
this was implemented via modifying the `dists` before the `layer_groups` are constructed.

```julia
# basic form that just always does what the user asked for
force_layers!(dists, forced_layers::Dict{Int, Int})
    for (node_id, target_layer) in forced_layers
        dists[node_id] = target_layer
    end
    return dists
end
```

Which we can fancy up to add the other features.
For example:

```julia
force_layers!(graphdists, forced_layers::Dict{Int, Int})
    # must process from end to beginning as otherwise can't move things after to make space
    ordered_forced_layers = sort(collect(forced_layers), by=last; rev=true)
    for (node_id, target_layer) in ordered_forced_layers
        curr_layer = dists[node_id]
        if target_layer < curr_layer
            @warn "Ignored opt_layer_assign for node $node_id; curr layer ($curr_layer) > desired layer ($target_layer)"
        elseif any(dists[child] <= target_layer for child in outneighbors(graph, node_id))
            @warn "Ignored opt_layer_assign for node $node_id; as placing it at $target_layer would place it on same layer, or later than it's children."
        else
            dists[node_id] = target_layer
        end
    end
    return dists
end
```

File: src/layering.jl [hunk 2]
```diff
@@ -1,9 +1,27 @@
 # This contains algorithms for breaking up a DAG into layers
-
-function layer_by_longest_path_to_source(graph)
+"Calculate the layer of each node"
+function layer_by_longest_path_to_source(graph, force_layer)
     dists = longest_paths(graph, sources(graph))
-    layer_groups = IterTools.groupby(i->dists[i], sort(vertices(graph), by=i->dists[i]))
-    return collect.(layer_groups)
+    force_layers!(graph, dists, force_layer
```
Comment (oxinabox) [inline src/layering.jl:5]:
```suggestion
    force_layers!(graph, dists, force_layer)
```

**Comment by davide-f** [comment]
@oxinabox Sorry for my long delay; I added the test case for the additional feature.
Now it should work as suggested including also the testing.
I will merge these changes in the other pull request

**Comment by davide-f** [comment]
Not sure why codecov fails, however the CI works perfectly

**Comment by davide-f** [comment]
The coverage seems to decrease because there are @warns that are not tested (clearly...), I think that the code should be fine

File: src/zarate.jl [hunk 3]
```diff
@@ -52,11 +58,12 @@ for e in edges(g)
 end
 ```
 """
-function solve_positions(layout::Zarate, original_graph)
+function solve_positions(layout::Zarate, original_graph;
+        force_layer = Vector{Pair{Int, Int}}())
```
Comment (oxinabox) [inline src/zarate.jl:62]:
```suggestion
function solve_positions(
    layout::Zarate, original_graph; force_layer = Vector{Pair{Int, Int}}()
)
```

**Comment by Warns** [comment]
> The coverage seems to decrease because there are @Warns that are not tested (clearly...), I think that the code should be fine

I think I was summoned here by mistake ^^

**Pull request merged**